File: src/java_version/arrays/ShellSort.java
package java_version.arrays;

public class ShellSort {

    public static void main(String[] args) {
        int[] array = Utils.arrayOfSevenElements();

        long startTime = System.nanoTime();

        for (int gap = array.length / 2; gap > 0; gap /= 2) {

            for (int i = gap; i < array.length; i++) {
                int newValue = array[i];

                int j = i;

                while (j >= gap && array[j - gap] > newValue) {
                    array[j] = array[j - gap];
                    j -= gap;
                }

                array[j] = newValue;
            }

        }

        System.out.print("Sorted Ended: " + (System.nanoTime() - startTime) + "\n");
        for (int i : array) {
            System.out.print("" + i + "\n");
        }
    }
}

File: src/kotlin_version/arrays/ShellSort.kt
package kotlin_version.arrays

fun main() {
    val startTime = System.nanoTime()

    var gap = arrayOfIntegers.size / 2

    while (gap > 0) {
        for (i in gap until arrayOfIntegers.size) {
            val newValue = arrayOfIntegers[i]
            var j = i

            while (j >= gap && arrayOfIntegers[j - gap] > 0) {
                arrayOfIntegers[j] = arrayOfIntegers[j - gap]
                j -= gap
            }
            arrayOfIntegers[j] = newValue
        }
        gap /= 2
    }

    val endTime = System.nanoTime()

    println("Execution time: ${endTime - startTime}")

    arrayOfIntegers.forEach {
        println(it)
    }
}